### ISSUE TITLE
Logging performance with early filtering

### ISSUE THREAD

**Issue by c42f** [JuliaLang/julia#30003]
Logging performance for events which are filtered early is several times worse than it can/should be. For example, as compared to julia-0.6 with `MicroLogging`, in 0.6 I have

```julia
julia> using MicroLogging

julia> function do_something_with_logger()
           for i in 1:10000
               @debug "print some info"
           end
       end

julia> @time do_something_with_logger()
  0.000533 seconds (4 allocations: 160 bytes)
```

which is somewhat respectable (though I hope we can do better in the future), whereas in 1.0.2 I find

```julia
julia> @time do_something_with_logger()
  0.003034 seconds (4 allocations: 160 bytes)
```

so we've suffered a 6x performance regression. Looking at the code I think this is likely due to the fact that we always execute `env_override_minlevel` (#25549) when we pass the first check. This is executed regardless of the value of the value of `logstate.min_enabled_level` making this cached value somewhat useless as a fast path.

A workaround is to use the global `disable_logging`, but this makes a unilateral decision for all other (unrelated) tasks which is not really ideal.

See also [this discusssion on discourse](https://discourse.julialang.org/t/logging-performance/17389), though in that case the author may be seeing the effect of https://github.com/JuliaLang/julia/issues/28400 which is now fixed.

**Comment by maleadt** [comment]
Just as a vote in favor of the behavior as introduced in https://github.com/JuliaLang/julia/pull/26265 (opposing eg. https://github.com/JuliaLang/julia/issues/28147#issuecomment-405694104): I find it extremely convenient to be able to `ENV["JULIA_DEBUG"] = "CUDAnative"` at run-time without having to restart Julia.

**Comment by c42f** [comment]
Yes, setting filtering via environment variable is useful functionality and addressed a real gap in functionality. But at the same time did it with an API (ie, environment variables) which is rather difficult to make fast. This issue is not to complain, but simply note that the design hasn't reached its potential yet: I'm convinced we can have both great performance and convenience at the same time.

I think it's quite difficult (impossible?) to get great performance by directly using environment variables. The reason is that as much filtering related information should be computed and cached as possible before you reach the site of the log event generation. With environment variables this isn't entirely possible because the environment could change at any time. So you end up reacting to those changes dynamically in a "pull based" fashion rather than setting them on the logger(s) in a "push based" fashion.

To solve this we "just" need a *convenient* declarative way of configuring loggers directly rather than indirectly using the environment. The reason this is hard is that defining a "convenient declarative configuration" tends to box in the design of the code which is being configured. To avoid that you really want a broad set of examples (of various implementations of `AbstractLogger`) on which to base the config system, and we just didn't have that in 0.7 so I admit I punted on it. I was hoping people would pursue it in packages for a while :-)

**Comment by oxinabox** [comment]
Discussed in https://github.com/oxinabox/LoggingExtras.jl/issues/20#issuecomment-545118059
We could change so that `ENV["JULIA_DEBUG"]` was a feature of the `ConsoleLogger`
which is only checked when the `ConsoleLogger` is instantiate,
and then that info can be stored in the `ConsoleLogger` object,
and used to define the `shouldlog` method.

It is kinda a breaking change, but it might not impact anyone.
The way it works right now is actively antithitical to how a compositional logging system,
or more broadly any kind of logger that is not monolythic, wants to work.
Since it is actually doing what `shouldlog` is supposed to do, but is bypassing that mechanism.

Longer term @c42f  and I have discussed that it might be nice to move towards the LoggingExtras.jl (or really the MicroLogging) style of pure sinks and compositional filters.
it is unclear how breakinmg that is, and LoggingExtras needs more time to ferment as a standalone package.
But it is possibly to define the compositional structure of filters + sinks that replicate the behavour of ` ENV["JULIA_DEBUG"]`.
